### ISSUE TITLE
Support Spark Generate with outer

### ISSUE THREAD

**Issue by boneanxs** [facebookincubator/velox#9630]
### Description

Spark supports `Generate` with an option outer to always include input even the unnested array/map is null or empty:

e.g.

```sql
-- without outer
spark-sql> select 1 as a, b from lateral view explode(array()) as  b;
Time taken: 0.127 seconds

-- with outer
spark-sql> select 1 as a, b from lateral view outer explode(array()) as  b;
1	NULL
Time taken: 0.152 seconds, Fetched 1 row(s)
```
The Spark community guide also describe this: https://spark.apache.org/docs/latest/sql-ref-syntax-qry-select-lateral-view.html

Since Gluten uses `Unnest` to support `Generate`(https://github.com/apache/incubator-gluten/pull/3068), but currently `Unnest` doesn't support outer and will ignore the input if the items in array/map is null/empty.   We can actually add an option `outer` inside `Unnest` to support it.

cc @zhouyuan

**Comment by rui-mo** [comment]
Hi @mbasmanova, would you take a look if this proposal makes sense? Thanks.

**Comment by zhouyuan** [comment]
CC: @marin-ma for her comments also as she did some checks on this feature

**Comment by mbasmanova** [comment]
@rui-mo @boneanxs This proposal sounds reasonable to me. One question. I remember Gluten was using ExpandNode, not UnnestNode. Would you clarify whether Gluten uses both ExpandNode and UnnestNode and if so how does it decide which one to use when?

**Comment by boneanxs** [comment]
> I remember Gluten was using ExpandNode, not UnnestNode. Would you clarify whether Gluten uses both ExpandNode and UnnestNode and if so how does it decide which one to use when?

Yes, Gluten uses both. Inside Spark, there're also 2 related operators, `GenerateExec` and `ExpandExec`, `GenerateExec` is usually used for the queries like `lateral view explode(array(items...))` to flat arrays and is replaced by `Unnest`, while `ExpandExec` is mainly for aggregation, used to compute `grouping set`, `count distinct()` etc, and is replaced by `Expand`.

**Comment by marin-ma** [comment]
@mbasmanova Spark's `GenerateExec` operator wraps one generator function. Unlike projection, the generator functions can produce arbitrary output rows for each input row based on their implementations. `explode` is one of the generator funciton (reference: https://spark.apache.org/docs/latest/api/sql/#explode) and there are other functions like `posexplode`, `inline`, `stack`, etc.

Currently in Gluten, each `GenerateExec` in Spark is converted into `UnnestNode` in the Velox pipeline. The key difference of Spark's `explode` and `UnnestNode` is that `explode` only accept one column as the input argument, while `UnnestNode` supports multiple columns. We also added support for `posexplode` and `inline` in Gluten by adding a post `ProjectionNode` after the `UnnestNode` to get the expected output.

e.g. Below SQL will be planned into 2 consecutive `GenerateExec`
```
scala> spark.sql("select c1, c2 from test lateral view explode(col1) as c1 lateral view explode(col2) as c2").explain
== Physical Plan ==
*(1) Generate explode(col2#16), [c1#17], false, [c2#18]
+- *(1) Generate explode(col1#15), [col2#16], false, [c1#17]
   +- *(1) Filter ((((size(col1#15, true) > 0) AND isnotnull(col1#15)) AND (size(col2#16, true) > 0)) AND isnotnull(col2#16))
      +- Scan hive spark_catalog.default.test [col1#15, col2#16], HiveTableRelation [`spark_catalog`.`default`.`test`, org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, Data Cols: [col1#15, col2#16], Partition Cols: []]
```
The converted Velox pipeline would be like
```
-- Unnest[n4_4] -> n4_3:INTEGER, C0:INTEGER
  -- Project[expressions: (n4_2:ARRAY<INTEGER>, "n2_3"), (n4_3:INTEGER, "C0"), (n4_4:ARRAY<INTEGER>, "n2_3")] -> n4_2:ARRAY<INTEGER>, n4_3:INTEGER, n4_4:ARRAY<INTEGER>
    -- Unnest[n2_4] -> n2_3:ARRAY<INTEGER>, C0:INTEGER
      -- Project[expressions: (n2_2:ARRAY<INTEGER>, "n0_0"), (n2_3:ARRAY<INTEGER>, "n0_1"), (n2_4:ARRAY<INTEGER>, "n0_0")] -> n2_2:ARRAY<INTEGER>, n2_3:ARRAY<INTEGER>, n2_4:ARRAY<INTEGER>
        -- Filter[expression: and(and(and(greaterthan(size("n0_0"),0),isnotnull("n0_0")),greaterthan(size("n0_1"),0)),isnotnull("n0_1"))] -> n0_0:ARRAY<INTEGER>, n0_1:ARRAY<INTEGER>
          -- ValueStream[] -> n0_0:ARRAY<INTEGER>, n0_1:ARRAY<INTEGER>
```

Initially, I was thinking about modifying the Spark's physical plan in Gluten by collapsing consecutive `GenerateExec` into a single operator and convert that operator into a single `UnnestNode`. However, this approach cannot cover all use cases in Spark. For example, a single `posexplode` function (https://spark.apache.org/docs/latest/api/sql/#posexplode)  is semantically equal to `UnnestNode` whose `withOrdinality_` is true, but 2 `posexplode` functions will generate 2 independent ordinality columns. In the case below, c1 is the ordinality column for col1, while c3 is the ordinality column for col2. We cannot get such output from a single `UnnestNode`.

```
scala> spark.sql("select * from test").show()
+------+------+
|  col1|  col2|
+------+------+
|[1, 2]|  NULL|
|  NULL|[3, 4]|
+------+------+

scala> spark.sql("select c1,c2,c3,c4 from test lateral view outer posexplode(col1) as c1,c2 lateral view outer posexplode(col2) as c3,c4").show()
+----+----+----+----+
|  c1|  c2|  c3|  c4|
+----+----+----+----+
|   0|   1|NULL|NULL|
|   1|   2|NULL|NULL|
|NULL|NULL|   0|   3|
|NULL|NULL|   1|   4|
+----+----+----+----+
```

To better support all generator functions and the outer semantic in SparkSQL, simply adding "null" can be insufficient and unnecessary. Spark can only process one column at a time with the `explode` function but handling multiple input columns with null much more complex in `Unnest` comparing with only one input column. Perhaps we should consider adding a new node, `GenerateNode`, in Velox, and implementing different generator functions independently.

**Comment by boneanxs** [comment]
@marin-ma Thanks for the detail explanation.

> Perhaps we should consider adding a new node, GenerateNode

Any special benefit for adding a new node? If I'm understanding correctly, if you want this new node only handles one input column to simplify logic, you still cannot reduce `GenerateExec` numbers if several consecutive `GenerateExec` nodes happens. 

> handling multiple input columns with null much more complex in Unnest comparing with only one input column

Multiple unnest cols won't add much complex to `Unnest` operator, We only need to check if all unnest cols are all null/empty of that row, if so output the `RepeatedColumns` once.

**Comment by marin-ma** [comment]
@boneanxs 
> Any special benefit for adding a new node?
The semantics of `Unnest` is strictly bound to the `explode` function in spark. However, there are other generator functions, like `posexplode`, `inline`, `stack`, etc. In Gluten, we use a workaround which adds a post projection after the `Unnest` for `posexplode` and `inline`. But for other generator functions, there's no workaround. The new node is used to call for different generator functions, to support other generator functions in sparksql, not only `explode`.

> you still cannot reduce GenerateExec numbers if several consecutive GenerateExec nodes happens.
That's true. The goal of the new node is to support more generator functions, not to reduce the numbers.

> Multiple unnest cols won't add much complex to Unnest operator
I'm not sure about this. But supporting multiple columns for spark is useless, because there will be only one column per each `UnnestNode`. In presto, Unnest accepts multiple columns, but there are no `outer` semantic, therefore it won't be needed by presto as well. https://prestodb.io/docs/current/sql/select.html#unnest

**Comment by mbasmanova** [comment]
@marin-ma Thank you for detailed explanation. Would you like to propose a new plan node design? I would be nice to describe all the generator functions you'd like to support.

**Comment by mbasmanova** [comment]
@xiaoxmeng is looking into adding this functionality in #13946 

CC: @oerling @rui-mo

**Comment by rui-mo** [comment]
There is a stale PR https://github.com/facebookincubator/velox/pull/9617 for this support from @boneanxs.

**Comment by marin-ma** [comment]
We found another solution in Gluten to address this comment https://github.com/facebookincubator/velox/issues/9630#issuecomment-2084695031

The only gap for the generator functions in Gluten was the support for the outer semantics in Velox. With https://github.com/facebookincubator/velox/pull/14095 merged, we should be able to support it now. Thanks @xiaoxmeng @mbasmanova 

Will get back to this when we finish supporting that in Gluten or if there are any further issues.

**Comment by mbasmanova** [comment]
@marin-ma Thank you for sharing this. Happy to hear https://github.com/facebookincubator/velox/pull/14095 might be helpful. Is there anything else needed to address this issue? If not, should we close it?

**Comment by marin-ma** [comment]
@mbasmanova The "generator with outer" semantics is now supported in Gluten, thanks to #14095 

Closing this one. Thanks!
